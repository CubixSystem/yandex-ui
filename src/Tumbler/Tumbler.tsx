import React, {
    FC,
    ChangeEventHandler,
    useCallback,
    useRef,
    RefObject,
    MouseEventHandler,
    useMemo,
    KeyboardEventHandler,
    KeyboardEvent,
    ReactNode,
    FocusEventHandler,
    useEffect,
} from 'react';
import { cn } from '@bem-react/classname';

import { Keys, isKeyCode } from '../lib/keyboard';
import { mergeRefs } from '../lib/mergeRefs';
import { useUniqId } from '../useUniqId';
import { TumblerLabel as Label } from './Label/Tumbler-Label';
import { TumblerButton as Button } from './Button/Tumbler-Button';
import { TumblerControl as Control } from './Control/Tumbler-Control';
import './Tumbler.css';

export const CHECKED_KEYS = [Keys.RIGHT, Keys.UP];
export const UNCHECKED_KEYS = [Keys.LEFT, Keys.DOWN];

export type TumblerProps = {
    /**
     * Состояние переключателя
     */
    checked: boolean;

    /**
     * Дополнительный класс у корневого DOM-элемента
     */
    className?: string;

    /**
     * Ссылка на нативный DOM-элемент нативного инпута
     */
    controlRef?: RefObject<HTMLInputElement>;

    /**
     * Неактивное состояние переключателя
     */
    disabled?: boolean;

    /**
     * Уникальный id компонента
     */
    id?: string;

    /**
     * Ссылка на корневой DOM-элемент компонента
     */
    innerRef?: RefObject<HTMLSpanElement>;

    /**
     * Подпись после переключателя
     */
    labelAfter?: ReactNode;

    /**
     * Подпись перед переключателем
     */
    labelBefore?: ReactNode;

    /**
     * Имя переключателя в форме
     */
    name?: string;

    /**
     * Обработчик, который вызывается при потере фокуса переключателем
     */
    onBlur?: FocusEventHandler<HTMLButtonElement>;

    /**
     * Обработчик, который вызывается при изменении состояния
     */
    onChange: ChangeEventHandler<HTMLInputElement>;

    /**
     * Обработчик, который вызывается при нажатии на переключатель
     */
    onClick?: MouseEventHandler<HTMLButtonElement>;

    /**
     * Обработчик, который вызывается при получении фокуса переключателем
     */
    onFocus?: FocusEventHandler<HTMLButtonElement>;

    /**
     * Обработчик, который вызывается при нажатии клавиши на клавиатуре (при этом переключатель должен быть в фокусе)
     */
    onKeyDown?: KeyboardEventHandler<HTMLButtonElement>;

    /**
     * Обработчик, который вызывается при отжатии клавиши на клавиатуре (при этом переключатель должен быть в фокусе)
     */
    onKeyUp?: KeyboardEventHandler<HTMLButtonElement>;

    /**
     * Целое число, определяющее должен ли переключатель участвовать
     * в последовательной навигации по всей странице с помощью клавиатуры
     */
    tabIndex?: number;

    /**
     * Всплывающая подсказка
     */
    title?: string;

    /**
     * Устанавливает фокус в компонент при монтировании
     */
    autoFocus?: boolean;

    /**
     * Устанавливает в компоненте обязательное состояние
     */
    required?: boolean;
};

export const cnTumbler = cn('Tumbler');

/**
 * Компонент, предназначенный для создания переключателя.
 *
 * @param {TumblerProps} props Свойства компонента.
 */
export const Tumbler: FC<TumblerProps> = ({
    checked,
    className,
    controlRef: htmlControlRef,
    disabled,
    // eslint-disable-next-line react-hooks/rules-of-hooks
    id = useUniqId(),
    innerRef,
    labelAfter,
    labelBefore,
    name,
    onBlur,
    onChange,
    onClick: htmlOnClick,
    onFocus,
    onKeyDown: htmlOnKeyDown,
    onKeyUp,
    tabIndex,
    title,
    autoFocus,
    required,
}) => {
    const buttonRef = useRef<HTMLButtonElement>(null);
    const controlRef = useRef<HTMLInputElement>(null);

    const controlId = `control-${id}`;
    const beforeLabelId = `before-label-${id}`;
    const afterLabelId = `after-label-${id}`;

    useEffect(() => {
        if (autoFocus && buttonRef.current !== null) {
            buttonRef.current.focus();
        }
        // TODO ISL-10952: разобраться почему не все необходимые переменные указаны в deps и убрать игнор
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, []);

    const labelledById = useMemo(() => {
        if (labelBefore && labelAfter) return checked ? beforeLabelId : afterLabelId;
        if (labelBefore) return beforeLabelId;
        if (labelAfter) return afterLabelId;
        return undefined;
    }, [checked, labelBefore, labelAfter, beforeLabelId, afterLabelId]);

    // prettier-ignore
    const onButtonClick = useCallback((event) => {
        if (controlRef.current !== null) {
            controlRef.current.click();
        }
        if (buttonRef.current !== null) {
            buttonRef.current.focus();
        }
        if (htmlOnClick !== undefined) {
            htmlOnClick(event);
        }
    }, [htmlOnClick]);

    // prettier-ignore
    const onLabelClick = useCallback(() => {
        if (buttonRef.current !== null) {
            buttonRef.current.focus();
        }
    }, []);

    // prettier-ignore
    const onKeyDown = useCallback((event: KeyboardEvent<HTMLButtonElement>) => {
        if (isKeyCode(event.keyCode, [Keys.ENTER, ...UNCHECKED_KEYS, ...CHECKED_KEYS])) {
            event.preventDefault();
        }
        const shouldForceChange =
            (checked && isKeyCode(event.keyCode, UNCHECKED_KEYS)) ||
            (!checked && isKeyCode(event.keyCode, CHECKED_KEYS));
        if (shouldForceChange && controlRef.current !== null) {
            controlRef.current.click();
        }
        if (htmlOnKeyDown !== undefined) {
            htmlOnKeyDown(event);
        }
    }, [checked, htmlOnKeyDown]);

    return (
        <span ref={innerRef} className={cnTumbler(null, [className])} id={id} title={title} aria-disabled={disabled}>
            {labelBefore && (
                <Label
                    disabled={Boolean(!checked && labelBefore && labelAfter)}
                    htmlFor={controlId}
                    id={beforeLabelId}
                    onClick={onLabelClick}
                >
                    {labelBefore}
                </Label>
            )}
            <Button
                tabIndex={disabled ? -1 : tabIndex}
                innerRef={buttonRef}
                checked={checked}
                onKeyDown={onKeyDown}
                onKeyUp={onKeyUp}
                onBlur={onBlur}
                onFocus={onFocus}
                onClick={onButtonClick}
                labelledBy={labelledById}
                required={required}
            />
            <Control
                id={controlId}
                name={name}
                checked={checked}
                innerRef={mergeRefs(controlRef, htmlControlRef)}
                onChange={onChange}
                required={required}
            />
            {labelAfter && (
                <Label
                    disabled={Boolean(checked && labelBefore && labelAfter)}
                    htmlFor={controlId}
                    id={afterLabelId}
                    onClick={onLabelClick}
                >
                    {labelAfter}
                </Label>
            )}
        </span>
    );
};

Tumbler.displayName = 'Tumbler';
