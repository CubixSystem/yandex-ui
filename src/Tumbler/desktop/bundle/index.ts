export * from '../../Tumbler.bundle/desktop';
