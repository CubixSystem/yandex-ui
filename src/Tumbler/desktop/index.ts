export * from '../Tumbler@desktop';
export * from '../_size/Tumbler_size_l';
export * from '../_size/Tumbler_size_m';
export * from '../_size/Tumbler_size_s';
export * from '../_view/Tumbler_view_default@desktop';
