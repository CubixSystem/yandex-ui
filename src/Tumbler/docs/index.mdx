---
id: Tumbler

path: /components/tumbler

title: Tumbler
description: Компонент, предназначенный для создания переключателя.

links:
 - label: source
   url: https://a.yandex-team.ru/arc_vcs/frontend/packages/lego-components/src/Tumbler
 - label: figma
   url: https://www.figma.com/file/CaoHFEO0w6PaE8cVaYFDoq8i/LEGO?node-id=411%3A129
 - label: storybook
   url: https://lego.yandex-team.ru/components/lego/latest/?storybook=%2Fstory%2Fcontrols-tumbler-desktop--playground

tabs:
 - Examples
 - Code
 - Modifiers
---
