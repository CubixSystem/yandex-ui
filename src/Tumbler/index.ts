export * from './Tumbler';
export * from './_size/Tumbler_size_l';
export * from './_size/Tumbler_size_m';
export * from './_size/Tumbler_size_s';
export * from './_view/Tumbler_view_default';
