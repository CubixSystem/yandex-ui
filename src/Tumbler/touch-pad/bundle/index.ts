import { compose, composeU } from '@bem-react/core';

import { Tumbler as TumblerTouchPad } from '../../Tumbler@touch-pad';
import { withSizeS } from '../../_size/Tumbler_size_s';
import { withSizeM } from '../../_size/Tumbler_size_m';
import { withSizeL } from '../../_size/Tumbler_size_l';
import { withViewDefault } from '../../_view/Tumbler_view_default';

export * from '../../Tumbler@touch-pad';

export const Tumbler = compose(
    composeU(withSizeS, withSizeM, withSizeL),
    withViewDefault,
)(TumblerTouchPad);
