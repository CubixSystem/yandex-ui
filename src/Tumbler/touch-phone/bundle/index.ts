export * from '../../Tumbler.bundle/touch-phone';
