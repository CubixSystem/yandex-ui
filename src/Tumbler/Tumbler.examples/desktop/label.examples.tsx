import React, { useState } from 'react';

import { Tumbler } from '@yandex-lego/components/Tumbler/desktop/bundle';

export const Label = () => {
    const [checked1, setChecked1] = useState(false);
    const [checked2, setChecked2] = useState(false);
    const [checked3, setChecked3] = useState(false);

    return (
        <>
            <div style={{ marginBottom: 8 }}>
                <Tumbler
                    size="m"
                    view="default"
                    checked={checked1}
                    onChange={() => setChecked1(!checked1)}
                    labelBefore="labelBefore"
                />
            </div>
            <div style={{ marginBottom: 8 }}>
                <Tumbler
                    size="m"
                    view="default"
                    checked={checked2}
                    onChange={() => setChecked2(!checked2)}
                    labelAfter="labelAfter"
                />
            </div>
            <div style={{ marginBottom: 8 }}>
                <Tumbler
                    size="m"
                    view="default"
                    checked={checked3}
                    onChange={() => setChecked3(!checked3)}
                    labelBefore={
                        <svg aria-label="labelBefore" width="16" height="16" xmlns="http://www.w3.org/2000/svg">
                            <path
                                d="M3.5 7.003a1.5 1.5 0 0 0-1.5 1.5v6a1.5 1.5 0 0 0 1.5 1.5h9a1.5 1.5 0 0 0 1.5-1.5v-6a1.5 1.5 0 0 0-1.5-1.5H12v-2a4 4 0 0 0-8 0v2h-.5zm2.5-2a2 2 0 1 1 4 0V7H6V5.003z"
                                fill="currentColor"
                            />
                        </svg>
                    }
                    labelAfter={
                        <svg aria-label="labelAfter" width="16" height="16" fill="none" xmlns="http://www.w3.org/2000/svg">
                            <path
                                d="M11 5.003a1 1 0 0 0 1-1 4 4 0 0 0-8 0v3h-.5a1.5 1.5 0 0 0-1.5 1.5v6a1.5 1.5 0 0 0 1.5 1.5h9a1.5 1.5 0 0 0 1.5-1.5v-6a1.5 1.5 0 0 0-1.5-1.5H6v-3a2 2 0 1 1 4 0 1 1 0 0 0 1 1z"
                                fill="currentColor"
                            />
                        </svg>
                    }
                />
            </div>
        </>
    );
};
