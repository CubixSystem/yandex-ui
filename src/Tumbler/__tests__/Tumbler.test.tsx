import React, { createRef } from 'react';

import { createClientRender, screen, fireEvent } from '../../internal/testing';
import { Tumbler as TumblerCommon } from '../Tumbler';
import { Tumbler as TumblerDesktop } from '../desktop';
import { Tumbler as TumblerTouchPad } from '../touch-pad';
import { Tumbler as TumblerTouchPhone } from '../touch-phone';

const platforms = [
    ['desktop', TumblerDesktop],
    ['touch-pad', TumblerTouchPad],
    ['touch-phone', TumblerTouchPhone],
] as const;

const noop = () => null;

describe.each(platforms)('Tumbler@%s', (_platform, Tumbler) => {
    const render = createClientRender();

    afterEach(() => {
        jest.useRealTimers();
    });

    test('должен вернуть полный шаблон компонента (snapshot)', () => {
        const { container } = render(
            <Tumbler
                checked={false}
                onChange={noop}
                tabIndex={1}
                name="switch"
                id="static-id"
                className="unit-mixin"
                labelBefore="labelBefore"
                labelAfter="labelAfter"
            />,
        );

        expect(container.firstChild).toMatchSnapshot();
    });

    test('должен устанавливать ссылку на корневой DOM элемент', () => {
        // NOTE: mergeRefs использует raf
        jest.useFakeTimers();
        const innerRef = createRef<HTMLSpanElement>();
        render(<Tumbler checked={false} onChange={noop} innerRef={innerRef} />);

        jest.runAllTimers();

        expect(innerRef.current).not.toBe(null);
    });

    test('должен устанавливать ссылку на нативный DOM элемент', () => {
        // NOTE: mergeRefs использует raf
        jest.useFakeTimers();
        const controlRef = createRef<HTMLInputElement>();
        render(<Tumbler checked={false} onChange={noop} controlRef={controlRef} />);

        jest.runAllTimers();

        expect(controlRef.current).not.toBe(null);
    });

    test('должен быть установлен корректный displayName', () => {
        expect(TumblerCommon.displayName).toBe('Tumbler');
    });

    test('должен устанавливать aria-pressed при изменении свойства checked', () => {
        const { setProps } = render(<Tumbler checked={false} onChange={noop} />);

        expect(screen.getByRole('button')).toHaveAttribute('aria-pressed', 'false');

        setProps({ checked: true });

        expect(screen.getByRole('button')).toHaveAttribute('aria-pressed', 'true');

        setProps({ checked: false });

        expect(screen.getByRole('button')).toHaveAttribute('aria-pressed', 'false');
    });

    test('должен устанавливать aria-disabled при изменении свойства disabled', () => {
        const { setProps } = render(<Tumbler title="element" disabled={false} checked={false} onChange={noop} />);

        expect(screen.getByTitle('element')).toHaveAttribute('aria-disabled', 'false');

        setProps({ disabled: true });

        expect(screen.getByTitle('element')).toHaveAttribute('aria-disabled', 'true');

        setProps({ disabled: false });

        expect(screen.getByTitle('element')).toHaveAttribute('aria-disabled', 'false');
    });

    test('должен устанавливать tabIndex=-1 при изменении свойства disabled', () => {
        const { setProps } = render(<Tumbler disabled={false} checked={false} onChange={noop} tabIndex={0} />);

        expect(screen.getByRole('button')).toHaveAttribute('tabindex', '0');

        setProps({ disabled: true });

        expect(screen.getByRole('button')).toHaveAttribute('tabindex', '-1');

        setProps({ disabled: false });

        expect(screen.getByRole('button')).toHaveAttribute('tabindex', '0');
    });

    test('должен вызывать обработчик onClick при нажатии на мышку', () => {
        const onClick = jest.fn();
        render(<Tumbler onClick={onClick} checked={false} onChange={noop} />);

        fireEvent.click(screen.getByRole('button'));

        expect(onClick).toHaveBeenCalledTimes(1);
    });

    test('должен вызывать обработчик onKeyDown при нажатии на клавиатуру', () => {
        const onKeyDown = jest.fn();
        render(<Tumbler onKeyDown={onKeyDown} checked={false} onChange={noop} />);

        fireEvent.keyDown(screen.getByRole('button'));

        expect(onKeyDown).toHaveBeenCalledTimes(1);
    });

    test('должен устанавливать фокус в кнопку после нажатия на лейбл либо кнопку', () => {
        render(
            <Tumbler checked={false} onChange={noop} labelBefore={<span>right</span>} labelAfter={<span>left</span>} />,
        );

        fireEvent.click(screen.getByRole('button'));

        expect(screen.getByRole('button')).toHaveFocus();

        // NOTE: Не используем click, так как react-testing-library
        // при клике на лейбл фокусирует инпут
        fireEvent.baseClick(screen.getByText('right'));

        expect(screen.getByRole('button')).toHaveFocus();

        fireEvent.baseClick(screen.getByText('left'));

        expect(screen.getByRole('button')).toHaveFocus();
    });
});
