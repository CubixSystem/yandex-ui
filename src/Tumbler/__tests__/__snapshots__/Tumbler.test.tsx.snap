// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tumbler@desktop должен вернуть полный шаблон компонента (snapshot) 1`] = `
<span
  class="Tumbler unit-mixin"
  id="static-id"
>
  <label
    aria-hidden="true"
    class="Tumbler-Label"
    id="before-label-static-id"
  >
    labelBefore
  </label>
  <button
    aria-labelledby="after-label-static-id"
    aria-pressed="false"
    class="Tumbler-Button"
    tabindex="1"
    type="button"
  />
  <input
    autocomplete="off"
    class="Tumbler-Control"
    hidden=""
    id="control-static-id"
    name="switch"
    type="checkbox"
    value="false"
  />
  <label
    aria-hidden="true"
    class="Tumbler-Label"
    for="control-static-id"
    id="after-label-static-id"
  >
    labelAfter
  </label>
</span>
`;

exports[`Tumbler@touch-pad должен вернуть полный шаблон компонента (snapshot) 1`] = `
<span
  class="Tumbler unit-mixin"
  id="static-id"
>
  <label
    aria-hidden="true"
    class="Tumbler-Label"
    id="before-label-static-id"
  >
    labelBefore
  </label>
  <button
    aria-labelledby="after-label-static-id"
    aria-pressed="false"
    class="Tumbler-Button"
    tabindex="1"
    type="button"
  />
  <input
    autocomplete="off"
    class="Tumbler-Control"
    hidden=""
    id="control-static-id"
    name="switch"
    type="checkbox"
    value="false"
  />
  <label
    aria-hidden="true"
    class="Tumbler-Label"
    for="control-static-id"
    id="after-label-static-id"
  >
    labelAfter
  </label>
</span>
`;

exports[`Tumbler@touch-phone должен вернуть полный шаблон компонента (snapshot) 1`] = `
<span
  class="Tumbler unit-mixin"
  id="static-id"
>
  <label
    aria-hidden="true"
    class="Tumbler-Label"
    id="before-label-static-id"
  >
    labelBefore
  </label>
  <button
    aria-labelledby="after-label-static-id"
    aria-pressed="false"
    class="Tumbler-Button"
    tabindex="1"
    type="button"
  />
  <input
    autocomplete="off"
    class="Tumbler-Control"
    hidden=""
    id="control-static-id"
    name="switch"
    type="checkbox"
    value="false"
  />
  <label
    aria-hidden="true"
    class="Tumbler-Label"
    for="control-static-id"
    id="after-label-static-id"
  >
    labelAfter
  </label>
</span>
`;
